runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,356 @@
+use {
+    crate::transaction_priority_details::GetTransactionPriorityDetails,
+    solana_measure::measure,
+    solana_sdk::{
+        clock::Slot,
+        pubkey::Pubkey,
+        saturating_add_assign,
+        transaction::{SanitizedTransaction, MAX_TX_ACCOUNT_LOCKS},
+    },
+    std::collections::HashMap,
+};
+
+#[derive(Debug, Default)]
+struct PrioritizationFeeMetrics {
+    // Count of writable accounts in slot
+    total_writable_accounts_count: u64,
+
+    // Count of writeable accounts with a minimum prioritization fee higher than the minimum transaction
+    // fee for this slot.
+    relevant_writable_accounts_count: u64,
+
+    // Total prioritization fees included in this slot.
+    total_prioritization_fee: u64,
+
+    // Accumulated time spent on tracking prioritization fee for each slot.
+    total_update_elapsed_us: u64,
+}
+
+impl PrioritizationFeeMetrics {
+    fn accumulate_total_prioritization_fee(&mut self, val: u64) {
+        saturating_add_assign!(self.total_prioritization_fee, val);
+    }
+
+    fn accumulate_total_update_elapsed_us(&mut self, val: u64) {
+        saturating_add_assign!(self.total_update_elapsed_us, val);
+    }
+
+    fn report(&self, slot: Slot) {
+        datapoint_info!(
+            "block_prioritization_fee",
+            ("slot", slot as i64, i64),
+            (
+                "total_writable_accounts_count",
+                self.total_writable_accounts_count as i64,
+                i64
+            ),
+            (
+                "relevant_writable_accounts_count",
+                self.relevant_writable_accounts_count as i64,
+                i64
+            ),
+            (
+                "total_prioritization_fee",
+                self.total_prioritization_fee as i64,
+                i64
+            ),
+            (
+                "total_update_elapsed_us",
+                self.total_update_elapsed_us as i64,
+                i64
+            ),
+        );
+    }
+}
+
+pub enum PrioritizationFeeError {
+    // Not able to get account locks from sanitized transaction, which is required to update block
+    // minimum fees.
+    FailGetTransactionAccountLocks,
+
+    // Not able to read priority details, including compute-unit price, from transaction.
+    // Compute-unit price is required to update block minimum fees.
+    FailGetTransactionPriorityDetails,
+
+    // Block is already finalized, trying to finalize it again is usually unexpected
+    BlockIsAlreadyFinalized,
+}
+
+/// Block minimum prioritization fee stats, includes the minimum prioritization fee of transactions in a
+/// block; and minimum fee for each writable accounts of all transactions in block. The only relevant
+/// write account minimum fees are those greater than block minimum fee, because the minimum fee needed to land
+/// a transaction is determined by Max( min_transaction_fee, min_writable_account_fees(key), ...)
```

Comment:
```suggestion
/// Block minimum prioritization fee stats, includes the minimum prioritization fee for a transaction in this
/// block; and the minimum fee for each writable account in all transactions in this block. The only relevant
/// write account minimum fees are those greater than the block minimum transaction fee, because the minimum fee needed to land
/// a transaction is determined by Max( min_transaction_fee, min_writable_account_fees(key), ...)
```